tiledb/sm/serialization/query.cc

```diff
@@ -2094,33 +2113,84 @@ Status query_serialize(
 
         // Concatenate buffers to end of message
         if (serialize_buffers) {
+          bool remote_global_order_write =
+              query->array()->is_remote() &&
+              query->type() == QueryType::WRITE &&
+              query->layout() == Layout::GLOBAL_ORDER;
+
           auto attr_buffer_builders = query_builder.getAttributeBufferHeaders();
           for (auto attr_buffer_builder : attr_buffer_builders) {
             const std::string name = attr_buffer_builder.getName().cStr();
+            auto buffer = query->buffer(name);
+            const auto& buffer_cache = query->get_remote_buffer_cache(name);
+
+            if (buffer.buffer_var_size_ != nullptr &&
+                buffer.buffer_var_ != nullptr) {
+              // Variable size buffers.
+              Buffer data;
+              Buffer var(buffer.buffer_var_, *buffer.buffer_var_size_);
+
+              // If we are not appending offsets we can use non-owning buffers.
+              if (buffer_cache.buffer_.size() > 0) {
+                Buffer prepend_data(
+                    buffer_cache.buffer_.data(), buffer_cache.buffer_.size());
+                Buffer prepend_var_data(
+                    buffer_cache.buffer_var_.data(),
+                    buffer_cache.buffer_var_.size());
+                RETURN_NOT_OK(
+                    serialized_buffer->add_buffer(std::move(prepend_data)));
+
+                // Ensure ascending order for appended user offsets.
+                // Copy user offsets so we don't modify buffer in client code.
+                throw_if_not_ok(
+                    data.write(buffer.buffer_, *buffer.buffer_size_));
+                for (uint64_t i = 0; i < data.size();
+                     i += constants::cell_var_offset_size) {
```

Comment:
Don't use c-style casts... Use static_cast instead. Also, cache buffer_cache.buffer_var_.size() outside the loop.